pkg/sql/lease.go

```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
 // inserted at the end of the lease set (i.e. it will be returned by
 // findNewest() from now on).
 //
+// acquireNodeLease begins an asynchronous process to acquire a lease.
+// Use errChan to receive any errors from the asynchronous process. If
+// error is not nil, expect the error channel to be nil.
+//
+// Use acquireWait to block until the lease is acquired.
+//
 // t.mu needs to be locked.
 func (t *tableState) acquireNodeLease(
 	ctx context.Context, m *LeaseManager, minExpirationTime hlc.Timestamp,
-) error {
+) (chan error, error) {
 	if m.isDraining() {
-		return errors.New("cannot acquire lease when draining")
+		return nil, errors.New("cannot acquire lease when draining")
 	}
 
-	// Notify when lease has been acquired.
+	// Notify when lease is being acquired.
 	t.mu.acquiring = make(chan struct{})
+	errChan := make(chan error)
+
+	err := t.stopper.RunAsyncTask(
+		ctx, "sql.tableState: async table lease acquisition", func(ctx context.Context) {
+			err := t.acquireNodeLeaseBlocking(ctx, m, minExpirationTime)
+			if err != nil {
+				log.Error(ctx, errors.Wrapf(err, "Async TableDescriptor lease acquisition error"))
+				errChan <- err
+			}
+			close(errChan)
+		})
+
+	if err != nil {
+		close(t.mu.acquiring)
+		close(errChan)
+		t.mu.acquiring = nil
+	}
+
+	return errChan, err
+}
+
+// acquireNodeLeaseBlocking does the blocking operation for acquiring a
+// table descriptor lease.
+//
+// t.mu.acquiring must be initialized before calling.
+func (t *tableState) acquireNodeLeaseBlocking(
+	ctx context.Context, m *LeaseManager, minExpirationTime hlc.Timestamp,
+) error {
+	defer t.mu.Unlock()
 	defer func() {
 		close(t.mu.acquiring)
 		t.mu.acquiring = nil
 	}()
```

Comment:
I think moving the close and setting to nil of acquiring is cleaner done in acquireNodeLease(). You already do that when you hit an error